sdk/vppagent/cross_connect.go

```diff
@@ -0,0 +1,105 @@
+package vppagent
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	l2 "github.com/ligato/vpp-agent/api/models/vpp/l2"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/local/connection"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/local/networkservice"
+	"github.com/networkservicemesh/networkservicemesh/sdk/common"
+	"github.com/networkservicemesh/networkservicemesh/sdk/endpoint"
+	"github.com/sirupsen/logrus"
+)
+
+// XConnect is a VPP Agent Cross Connect composite
+type XConnect struct {
+	endpoint.BaseCompositeEndpoint
+	Workspace   string
+	Connections map[string]*ConnectionData
+}
+
+// Request implements the request handler
+func (xc *XConnect) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
+
+	if xc.GetNext() == nil {
+		logrus.Fatal("The VPP Agent Cross Connect composite requires that there is Next set")
+	}
+
+	incomingConnection, err := xc.GetNext().Request(ctx, request)
+	if err != nil {
+		logrus.Errorf("Next request failed: %v", err)
+		return nil, err
+	}
+
+	opaque := xc.GetNext().GetOpaque(incomingConnection)
+	if opaque == nil {
+		err := fmt.Errorf("received empty data from Next")
+		logrus.Errorf("Unable to find connection data: %v", err)
+		return nil, err
+	}
+	connectionData := opaque.(*ConnectionData)
+
+	if connectionData.DstName == "" {
+		err := fmt.Errorf("found empty destination name")
+		logrus.Errorf("Invalid connection data: %v", err)
+		return nil, err
+	}
+	if connectionData.SrcName == "" {
```

Comment:
Src and Dst come from our forwarding plane concepts. Maybe we can reconsider these - In and Out or something esle.

Reply:
In ConnectionContext type we have SrcIpAddr and DstIpAddr fields. D[stIpAddr is used in MemifConnect](https://github.com/networkservicemesh/networkservicemesh/pull/1303/files#diff-c69e69e1dbbcae3efe6b8d16e38e8bd2R59) to decide whether ip address should be assigned.

Won't it lead to ambiguity if we will work with both src/dst and in/out?

Reply:
OK so in cross connects we do not have IPs, that's the point of the cross connect - transparrent passhtrough. No one in the shain will know that the Corssconnect even exists. Both interfaces in the cross connect do not have IPs.

But that is a separate story. All these small composites should be completely independent. Or at least very loosely coupled. So whatever you name things here shoudl not be directly related to what other compsoites name things.

Src and Dst are not accurate here, since this is a directionless corssconnec. Even SideA and SideB is more direction independent and just says - htere are two sides here and I am connecting them, not really caring where they come from.
Hope that makes sense.

Reply:
Thank you a lot for the clarification, much clearer for me now. Will replace with In/Out because ACL needs to be applied only for incoming connection, so I cannot use smth like SideA/SideB or a slice of names.

Reply:
Done